### DISCUSSION TITLE
Add the ability to set the segment in the PublishedRouter

### DISCUSSION THREAD

**Discussion by drpeck** [umbraco/Umbraco-CMS#13594] [Features and ideas]
**As things stand**
The `PublishedRouter` handles creation of the `VariationContext`. It sets the `Culture` property with a value supplied by the `IPublishedRequestBuilder`. [No segment is set](https://github.com/umbraco/Umbraco-CMS/blob/46414f2c646d9a04e2d3614876a7a311cba8653a/src/Umbraco.Core/Routing/PublishedRouter.cs#L181).

The `IPublishedRequestBuilder` has a `SetCulture` ([link](https://github.com/umbraco/Umbraco-CMS/blob/b69afe81f3f6fcd37480b3b0295a62af44ede245/src/Umbraco.Core/Routing/IPublishedRequestBuilder.cs#L72)), the value of which is used for setting the culture on the `VariationContext` ([link](https://github.com/umbraco/Umbraco-CMS/blob/46414f2c646d9a04e2d3614876a7a311cba8653a/src/Umbraco.Core/Routing/PublishedRouter.cs#L226)). This method might be called by a content finder for example. There is no such comparable `SetSegment()` method.

**Proposal**
Would you support a PR that adds a `SetSegment()` to the  IPublishedRequestBuilder, and also modifies `PublishedRouter` to use this set value when creating the `new VariationContext()`?

AFAIK the best current approach to set the `Segment` property on the `VariationContext` is by using a `RoutingRequestNotification`. In my case I'm identifying the segment in a `IContentFinder` (based on the URL) and then having to calculate the same value again later in a `RoutingRequestNotification` to set the `VariationContent`. This is because the `VariationContext` is [overwritten](            SetVariationContext(builder.Culture);) after the content finders have run. This double handling could be avoided with this approach.

**Comment by drpeck** [comment]
This commit should be the totality of the change:
https://github.com/drpeck/Umbraco-CMS/commit/5d8a2718a7a2e1990fbefdd3b87d135fdbdafc56

**Comment by lassefredslund** [comment]
Hi @drpeck 
Thanks for the idea and the PR.
To me, this sounds OK, however, I know that @PerplexDaniel uses segments a lot, so maybe he could verify it too?

**Reply by PerplexDaniel** [reply]
@lassefredslund 
We use segments in uMarketingSuite through a custom `IVariationContextAccessor`, we do not alter it through the PublishedRouter. 
This change seems fine and should not break anything as far as I can tell. 
This does seem like a breaking change though as it modifies some public interfaces so it might take a while before Umbraco can release this (i.e., earliest in v12?) but that's for HQ to decide.

**Reply by drpeck** [reply]
A custom `IVariationContextAccessor` is a clever idea (yoink).

I'm not sure what the definition of a breaking change is. Public interfaces change, but only with new members. The only modified existing member is a private method. I hadn't anticipated that this would be classified as a breaking change, but I suppose it could upset someone using reflection.

**Reply by nul800sebastiaan** [reply]
Mostly upsets people who have implemented the interface 😉

But it can be fixed by default implementation (see: https://github.com/umbraco/Umbraco-CMS/pull/12218#issuecomment-1089981684 and onwards).

**Reply by PerplexDaniel** [reply]
@drpeck You might be right they do not consider new members a breaking change, but I would think if I currently have some custom implementation of `IPublishedRequestBuilder` in my project (not sure why, but let's imagine) and then upgrade Umbraco which includes your commit then suddenly my code will not compile anymore since I have not specified the `string? Segment` property in my class. So build should fail with an exception like `MyCustomPublishedRequestBuilder does not implement interface member IPublishedRequestBuilder.Segment` (and also for the `SetSegment` method).

**Reply by nul800sebastiaan** [reply]
Ah sorry.. that was a new member on the interface, yeah that's hard to avoid. In the past we've then made a `IPublishedRequestBuilder2` that inherits from `IPublishedRequestBuilder` and used that. It's ugly but it works.

**Reply by drpeck** [reply]
Fair point(s).

I can absolutely create ` IPublishedRequestBuilder2` but that's going to create some technical debt. Might this be better pushed to v12 as it's not causing anyone any issues right now? @PerplexDaniel 's workaround solves my issue for V11.

**Reply by nul800sebastiaan** [reply]
I agree, best to target v12 as there's a great workaround for now! 👍  (`v12/dev` branch)

**Reply by drpeck** [reply]
[PR submitted](https://github.com/umbraco/Umbraco-CMS/pull/13927) but checks are proving to be an issue at the time of writing for some reason.
